data/finset/basic.lean

```diff
@@ -0,0 +1,704 @@
+@finset.induction_on _ _ (λ x, x ≠ empty → ∃ a : α, mem a x) s 
+(λ h, absurd rfl h)
+(by intros a s nin ih h; existsi a; apply mem_insert)
+
+theorem eq_empty_of_card_eq_zero {s : finset α} (H : card s = 0) : s = ∅ :=
+@finset.induction_on _ _ (λ x, card x = 0 → x = empty) s 
+(λ h, rfl) 
+(by intros a s' H1 Ih H; rw (card_insert_of_not_mem H1) at H; contradiction) H
+
+end insert
+
+/- erase -/
+section erase
+variable [h : decidable_eq α]
+include h
+
+def erase (a : α) (s : finset α) : finset α :=
+quot.lift_on s
+  (λ l, to_finset_of_nodup (l.1.erase a) (nodup_erase_of_nodup a l.2))
+  (λ (l₁ l₂ : nodup_list α) (p : l₁ ~ l₂), quot.sound (perm.erase_perm_erase_of_perm a p))
+
+theorem not_mem_erase (a : α) (s : finset α) : a ∉ erase a s :=
+quot.induction_on s
+  (λ l, list.mem_erase_of_nodup _ l.2)
+
+theorem card_erase_of_mem {a : α} {s : finset α} : a ∈ s → card (erase a s) = pred (card s) :=
+quot.induction_on s (λ l ainl, list.length_erase_of_mem ainl)
+
+theorem card_erase_of_not_mem {a : α} {s : finset α} : a ∉ s → card (erase a s) = card s :=
+quot.induction_on s (λ l nainl, length_erase_of_not_mem nainl)
+
+theorem erase_empty (a : α) : erase a ∅ = ∅ := rfl
+
+theorem ne_of_mem_erase {a b : α} {s : finset α} : b ∈ erase a s → b ≠ a :=
+by intros h beqa; subst b; exact absurd h (not_mem_erase _ _)
+
+theorem mem_of_mem_erase {a b : α} {s : finset α} : b ∈ erase a s → b ∈ s :=
+quot.induction_on s (λ l bin, mem_of_mem_erase bin)
+
+theorem mem_erase_of_ne_of_mem {a b : α} {s : finset α} : a ≠ b → a ∈ s → a ∈ erase b s :=
+quot.induction_on s (λ l n ain, list.mem_erase_of_ne_of_mem n ain)
+
+theorem mem_erase_iff (a b : α) (s : finset α) : a ∈ erase b s ↔ a ∈ s ∧ a ≠ b :=
+iff.intro
+  (λ H, and.intro (mem_of_mem_erase H) (ne_of_mem_erase H))
+  (λ H, mem_erase_of_ne_of_mem (and.right H) (and.left H))
+
+theorem mem_erase_eq (a b : α) (s : finset α) : a ∈ erase b s = (a ∈ s ∧ a ≠ b) :=
+propext (mem_erase_iff _ _ _)
+
+open decidable
+theorem erase_insert {a : α} {s : finset α} : a ∉ s → erase a (insert a s) = s :=
+λ anins, finset.ext (λ b, by_cases
+  (λ beqa : b = a, iff.intro
+    (λ bin, by subst b; exact absurd bin (not_mem_erase _ _))
+    (λ bin, by subst b; contradiction))
+  (λ bnea : b ≠ a, iff.intro
+    (λ bin,
+       have b ∈ insert a s, from mem_of_mem_erase bin,
+       mem_of_mem_insert_of_ne this bnea)
+    (λ bin,
+      have b ∈ insert a s, from mem_insert_of_mem _ bin,
+      mem_erase_of_ne_of_mem bnea this)))
+
+theorem insert_erase {a : α} {s : finset α} : a ∈ s → insert a (erase a s) = s :=
+λ ains, finset.ext (λ b, by_cases
+  (λ h : b = a, iff.intro
+    (λ bin, by subst b; assumption)
+    (λ bin, by subst b; apply mem_insert))
+  (λ hn : b ≠ a, iff.intro
+    (λ bin, mem_of_mem_erase (mem_of_mem_insert_of_ne bin hn))
+(λ bin, mem_insert_of_mem _ (mem_erase_of_ne_of_mem hn bin))))
+end erase
+
+/- union -/
+section union
+variable [h : decidable_eq α]
+include h
+
+def union (s₁ s₂ : finset α) : finset α :=
```

Comment:
`has_union`

Reply:
By the way, here and in the other places, you can just write `[decidable_eq α]` and leave out the `include`.

Reply:
Fixed, thanks!